crates/bevy_tasks/src/task.rs

```diff
@@ -36,25 +77,134 @@ impl<T> Task<T> {
     /// While it's possible to simply drop the [`Task`] to cancel it, this is a cleaner way of
     /// canceling because it also waits for the task to stop running.
     ///
-    /// See `async_executor::Task::cancel`
+    /// Canceling tasks is unsupported on the web, and this is the same as awaiting the task.
     pub async fn cancel(self) -> Option<T> {
-        self.0.cancel().await
+        cfg::web! {
+            if {
+                // Await the task and handle any panics.
+                match self.0.recv().await {
+                    Ok(Ok(value)) => Some(value),
+                    Err(_) => None,
+                    Ok(Err(panic)) => {
+                        // drop this to prevent the panic payload from resuming the panic on drop.
+                        // this also leaks the box but I'm not sure how to avoid that
+                        core::mem::forget(panic);
+                        None
+                    }
+                }
+            } else {
+                // Wait for the task to become canceled
+                self.0.cancel().await
+            }
+        }
     }
 
     /// Returns `true` if the current task is finished.
     ///
-    ///
     /// Unlike poll, it doesn't resolve the final value, it just checks if the task has finished.
     /// Note that in a multithreaded environment, this task can be finished immediately after calling this function.
     pub fn is_finished(&self) -> bool {
-        self.0.is_finished()
+        cfg::web! {
+            if {
+                // We treat the task as unfinished until the result is sent over the channel.
+                !self.0.is_empty()
+            } else {
+                // Defer to the `async_task` implementation.
+                self.0.is_finished()
+            }
+        }
     }
 }
 
+// Tasks are futures that can be polled
```

Comment:
All futures can be polled. Not sure what this comment is for.

Reply:
Yeah no idea what I was thinking there. I'll remove that.